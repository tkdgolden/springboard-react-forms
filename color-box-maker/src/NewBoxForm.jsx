import { useState } from 'react'


const NewBoxForm = ({ addBox }) => {
    const [fData, setFormData] = useState({color: "", height: "", width: ""});
    const INITIAL_STATE = {color: "", height: "", width: ""};

    const handleChange = evt => {
        const { name, value } = evt.target;

        setFormData(fData => ({
            ...fData,
            [name]: value
        }))
    }
    
    const handleSubmit = evt => {
        evt.preventDefault();
        if (fData.color && fData.height && fData.width) {
            addBox(fData);
            setFormData(INITIAL_STATE);
        }
    }

    return (
        <form onSubmit={ handleSubmit }>
            <label htmlFor="color">Color: </label>
            <input 
                id="color" 
                type="text" 
                placeholder="color" 
                value={fData.color} 
                onChange={handleChange} 
                name="color" 
            />
            <label htmlFor="height">Height: </label>
            <input
                id="height" 
                type="number" 
                placeholder="height" 
                value={fData.height} 
                onChange={handleChange} 
                name="height" 
            />
            <label htmlFor="width">Width: </label>
            <input 
                id="width"
                type="number" 
                placeholder="width" 
                value={fData.width} 
                onChange={handleChange} 
                name="width" 
            />
            <button>Add!</button>
        </form>
    );
};

export default NewBoxForm