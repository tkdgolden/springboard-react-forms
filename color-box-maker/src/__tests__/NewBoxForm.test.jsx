import { describe, expect, test } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import NewBoxForm from "../NewBoxForm";
import { useState } from "react";
import * as matchers from '@testing-library/jest-dom/matchers'

console.log(matchers);

describe("NewBoxForm", () => {
  // smoketest
  test("renders", () => {
    render(<NewBoxForm />);
    const title = screen.getByText("color", { exact: false });
    expect(title.textContent).toBe("Color: ");
  });

  // snapshot
  test("hasn't changed", () => {
    const {asFragment} = render(<NewBoxForm />);
    expect(asFragment()).toMatchSnapshot();
  });


  it("can input color", function() {
    const { getByLabelText, queryByText } = render(<NewBoxForm />);
  
    // default
    const colorInput = getByLabelText("Color:");
    expect(colorInput.value).toBe("");
  
    // fill out the form
    fireEvent.change(colorInput, { target: { value: "blue" }});
  
    // changed
    expect(colorInput.value).toBe("blue");
  });

  it("can input height", function() {
    const { getByLabelText, queryByText } = render(<NewBoxForm />);
  
    // default
    const heightInput = getByLabelText("Height:");
    expect(heightInput.value).toBe("");
  
    // fill out the form
    fireEvent.change(heightInput, { target: { value: "120" }});
  
    // changed
    expect(heightInput.value).toBe("120");
  });

  it("can input width", function() {
    const { getByLabelText, queryByText } = render(<NewBoxForm />);
  
    // default
    const widthInput = getByLabelText("Width:");
    expect(widthInput.value).toBe("");
  
    // fill out the form
    fireEvent.change(widthInput, { target: { value: "80" }});
  
    // changed
    expect(widthInput.value).toBe("80");
  });


});