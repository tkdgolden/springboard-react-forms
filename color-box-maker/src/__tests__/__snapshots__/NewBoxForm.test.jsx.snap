// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewBoxForm > hasn't changed 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color: 
    </label>
    <input
      id="color"
      name="color"
      placeholder="color"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      type="number"
      value=""
    />
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      type="number"
      value=""
    />
    <button>
      Add!
    </button>
  </form>
</DocumentFragment>
`;

exports[`NewBoxForm > renders 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color: 
    </label>
    <input
      id="color"
      name="color"
      placeholder="color"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      type="number"
      value=""
    />
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      type="number"
      value=""
    />
    <button>
      Add!
    </button>
  </form>
</DocumentFragment>
`;
